xml-light2/xml.ml

```diff
@@ -19,189 +19,186 @@
-		Element (name_local, attrs', List.filter (fun xml -> not (is_empty xml)) children)
-		in
-	let data s = PCData s in
-	match Xmlm.peek i with
-	| `Dtd _ -> snd (Xmlm.input_doc_tree ~el ~data i)
-	| _      -> Xmlm.input_tree ~el ~data i
+  let el (tag: Xmlm.tag) (children: xml list) : xml =
+    let name_local = snd (fst tag) in
+    let attrs' = List.map (fun (nameattr, str) -> (snd nameattr, str)) (snd tag) in
+    Element (name_local, attrs', List.filter (fun xml -> not (is_empty xml)) children)
+  in
+  let data s = PCData s in
+  match Xmlm.peek i with
+  | `Dtd _ -> snd (Xmlm.input_doc_tree ~el ~data i)
+  | _      -> Xmlm.input_tree ~el ~data i
 
 let parse i =
-	try _parse i
-	with
-	| Xmlm.Error ((line, col), msg) ->
-		let pos = {
-			eline = line; eline_start = line;
-			emin = col; emax = col
-		} in
-		let err = Xmlm.error_message msg in
-		raise (Error (err, pos))
+  try _parse i
+  with
+  | Xmlm.Error ((line, col), msg) ->
+    let pos = {
+      eline = line; eline_start = line;
+      emin = col; emax = col
+    } in
+    let err = Xmlm.error_message msg in
+    raise (Error (err, pos))
 
 (* common parse function *)
 let parse_file file =
-	let chan = open_in file in
-	try
-		let i = Xmlm.make_input (`Channel chan) in
-		let ret = parse i in
-		close_in chan;
-		ret
-	with exn ->
-		close_in_noerr chan; raise exn
+  let chan = open_in file in
+  try
+    let i = Xmlm.make_input (`Channel chan) in
+    let ret = parse i in
+    close_in chan;
+    ret
+  with exn ->
+    close_in_noerr chan; raise exn
 
 let parse_in chan =
-	let i = Xmlm.make_input (`Channel chan) in
-	parse i
+  let i = Xmlm.make_input (`Channel chan) in
+  parse i
 
 let parse_string s =
-	let i = Xmlm.make_input (`String (0, s)) in
-	parse i
+  let i = Xmlm.make_input (`String (0, s)) in
+  parse i
 
 let parse_bigbuffer b =
-	let n = ref Int64.zero in
-	let aux () =
-		try 
-			let c = Stdext.Bigbuffer.get b !n in
-			n := Int64.add !n Int64.one;
-			int_of_char c
-		with _ -> raise End_of_file in
-	let i = Xmlm.make_input (`Fun aux) in
-	parse i
-
-(* common output function *)
-let substitute list s =
-	s
+  let n = ref Int64.zero in
+  let aux () =
+    try 
+      let c = Xapi_stdext_bigbuffer.Bigbuffer.get b !n in
+      n := Int64.add !n Int64.one;
+      int_of_char c
+    with _ -> raise End_of_file in
+  let i = Xmlm.make_input (`Fun aux) in
+  parse i
 
 let esc_pcdata data =
-	let buf = Buffer.create (String.length data + 10) in
-	for i = 0 to String.length data - 1
-	do
-		let s = match data.[i] with
-		| '>'    -> "&gt;";
-		| '<'    -> "&lt;";
-		| '&'    -> "&amp;";
-		| '"'    -> "&quot;";
-		| c when (c >= '\x20' && c <= '\xff')
-		      || c = '\x09' || c = '\x0a' || c = '\x0d'
-		         -> String.make 1 c
-		| _      -> ""
-			in
-		Buffer.add_string buf s
-	done;
-	Buffer.contents buf
+  let buf = Buffer.create (String.length data + 10) in
+  for i = 0 to String.length data - 1
```

Comment:
`String.iter` would be more appropriate.

Reply:
True, or it could be sped up as a recusive lookup that exits at the first `false`.